daemon/migrator.c

```diff
@@ -0,0 +1,788 @@
+}
+
+static const gchar *
+approval_profile(const approval_t *self)
+{
+    /* Application profile path, i.e. application's desktop entry file path */
+    return self->apr_profile;
+}
+
+static uid_t
+approval_uid(const approval_t *self)
+{
+    /* UID from approval file path */
+    return self->apr_uid;
+}
+
+static stringset_t *
+approval_permissions(const approval_t *self)
+{
+    /* Approved permissions */
+    return self->apr_permissions;
+}
+
+static const gchar *
+approval_organization(const approval_t *self)
+{
+    /* The same as OrganizationName in Sailjail section */
+    return self->apr_organization;
+}
+
+static const gchar *
+approval_application(const approval_t *self)
+{
+    /* The same as ApplicationName in Sailjail section */
+    return self->apr_application;
+}
+
+static const migrator_t *
+approval_migrator(const approval_t *self)
+{
+    return self->apr_migrator;
+}
+
+static const control_t *
+approval_control(const approval_t *self)
+{
+    return migrator_control(approval_migrator(self));
+}
+
+/* ------------------------------------------------------------------------- *
+ * UTILITY
+ * ------------------------------------------------------------------------- */
+
+static void
+remove_approval_path(const gchar *path)
+{
+    /* Remove path and any empty parent directories until HOMESCREEN_DATA_PATH */
+    if( unlink(path) == -1 ) {
+        log_err("%s could not remove: %m", path);
+    }
+    else {
+        gchar *dir_path = path_dirname(path);
+        while( may_remove_approval_path(dir_path) && rmdir(dir_path) == 0 ) {
+            gchar *temp = dir_path;
+            dir_path = path_dirname(temp);
+            g_free(temp);
+        }
+        log_debug("%s cleaned up to %s", path, dir_path);
+        g_free(dir_path);
+    }
+}
+
+static bool
+may_remove_approval_path(const gchar *path)
+{
+    path = without_leading_data_path(path);
+    return path && *path;
+}
+
+static gchar *
+profile_from_approval_path(const gchar *path)
+{
+    /* Returns application profile path from approval file path, i.e.
+     * "/var/lib/sailjail-homescreen/UID/usr/share/applications/Y.desktop/X-Sailjail"
+     * returns "/usr/share/applications/Y.desktop".
+     */
+    path = without_leading_data_path(path);
+    if( path )
+        path = strchr(path, '/');
+    return path_dirname(path);
+}
+
+static uid_t
+uid_from_approval_path(const gchar *path, const control_t *control)
+{
+    gint64 uid = MIGRATION_UID_UNDEFINED;
+    path = without_leading_data_path(path);
+    if( path ) {
+        uid = g_ascii_strtoll(path, NULL, 10);
+    if( uid < control_min_user(control) ||
+        uid > control_max_user(control) )
+        return MIGRATION_UID_UNDEFINED;
+    }
+    return (uid_t)uid;
+}
+
+static const gchar *
+without_leading_data_path(const gchar *path)
+{
+    if( path && g_str_has_prefix(path, HOMESCREEN_DATA_PATH) )
+        return path + strlen(HOMESCREEN_DATA_PATH);
```

Comment:
All other case of \#define FOO_PATH omit trailing "/" -> this *looks* like it might stumble like:

- /var/lib/sailjail-homescreenfoo/bar -> foo/bar

But as it is \#define HOMESCREEN_DATA_PATH "/var/lib/sailjail-homescreen/" and that slash is included in strlen result -> that trap is avoided, but to me it looks like it susceptible to skip-too-much problem:

- /var/lib/sailjail-homescreen/foo/bar -> foo/bar

Dropping the trailing slash from macro and inserting it in here might be clearer?

    #define HOMESCREEN_DATA_PATH "/var/lib/sailjail-homescreen"

    if( path && g_str_has_prefix(path, HOMESCREEN_DATA_PATH "/") )
        return path + strlen(HOMESCREEN_DATA_PATH);

(with only slight case of "strncmp way would avoid effort and memory use doubling" tingling somewhere behind ear)

Reply:
It was correct already AFAICT. I added a comment what it is doing which should help understanding. I also swapped it to use  strncmp instead and made that "/" although it's needed everywhere that this define is used at but at least it's more consistent.